### PR TITLE
Add finalizers for specialresources add  metrics.go framework and two basic metrics

### PR DESCRIPTION
**Pull request by dagrayvid**
This PR adds finalizers to specialresources and metrics to SRO.
The implementation of finalizers is based heavily on the [Operator SDK example](https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#handle-cleanup-on-deletion).

The initial metrics added are:
 - Number of specialresources being managed
 - Which states have been completed per specialresource
 
More metrics can be added in the future. See metrics.go

Currently the only cleanup step done before finalizers are reconciled on specialresource deletion is to clear the completed states metric, however the scaffolding is there to later potentially add more complex cleanup later if necessary. We may want to later add some way to specify custom cleanup steps to the CRD.

### CODE REVIEW

File: controllers/specialresource.go [hunk 1]
```diff
@@ -121,10 +129,39 @@ func ReconcilerSpecialResources(r *SpecialResourceReconciler, req ctrl.Request)
 			}
 		}
 
-		//log = r.Log.WithValues("specialresource", r.parent.Name)
-		log = r.Log.WithName(color.Print(r.parent.Name, color.Green))
+		log = r.Log.WithName(color.Print(r.specialresource.Name, color.Green))
 		log.Info("Reconciling")
 
+		isMarkedToBeDeleted := r.specialresource.GetDeletionTimestamp() != nil
```
Comment (zvonkok) [inline controllers/specialresource.go:133]: Can we move this whole block into a function an add it to `finalizers.go` file?

Comment (zvonkok) [reply controllers/specialresource.go:133]: e.g. ReconcileFinalizers(r)

File: controllers/specialresource.go [hunk 2]
```diff
@@ -181,3 +218,57 @@ func createSpecialResourceFrom(r *SpecialResourceReconciler, name string) (srov1
 
 	return specialresource, errs.New("Created new SpecialResource we need to Reconcile")
 }
+
+func (r *SpecialResourceReconciler) finalizeSpecialResource(toFinalize srov1beta1.SpecialResource) error {
```
Comment (zvonkok) [inline controllers/specialresource.go]: As well as this function to `finalizers.go`

File: controllers/specialresource.go (hunk 1 above)
Comment (dagrayvid) [reply controllers/specialresource.go:133]: Yep, good idea

**Comment by ArangoGutierrez** [comment]
Is this ready for review @dagrayvid ? should we unhold?

**Comment by dagrayvid** [comment]
> Is this ready for review @dagrayvid ? should we unhold?

@ArangoGutierrez  I don't expect any major changes so should be ready for review, but still trying to investigate some weird behavior with the reconciliation loop before unholding

**Comment by ArangoGutierrez** [comment]
This PR is also adding finalizers, could to document that in the PR description so it is properly stored on the merge commit

**Pull request merged**